File: Runtime/Cpp/StateMachine/state_base.cpp
// Copyright ax, Inc. All Rights Reserved.

#include "ax-fsm/state_base.h"
#include "ax-fsm/exceptions.h"

using namespace std;

namespace ax { namespace fsm
{
    StateBase::StateBase(StateToken* token)
    {
        if (token == nullptr)
            throw new invalid_argument("Invalid 'token' argument.");

        _stateManager = nullptr;
        _token = token;
    }

    const StateToken* const StateBase::GetToken() const
    {
        return _token;
    }

    StateManager* StateBase::GetStateManager()
    {
        return _stateManager;
    }

    void* StateBase::GetContext()
    {
        if (_stateManager == nullptr)
            return nullptr;

        return _stateManager->GetContext();
    }

    void StateBase::Initialize(StateManager* parent)
    {
        _stateManager = parent;
        OnInitialized();
    }

    void StateBase::OnInitialized() { }

    void StateBase::OnEnter(StateEnterEventArgs* e) { }
    void StateBase::OnExit(StateExitEventArgs* e) { }

    void StateBase::RegisterActionHandler(ActionToken* action, StateHandler handler)
    {
        if (_handlers.find(action) != _handlers.end())
            throw new exception("Action already registered.");

        _handlers[action] = handler;
    }

    void StateBase::Handle(ActionToken* action, StateData* data, TransitionInfo* result)
    {
        if (action == nullptr)
            throw new invalid_argument("Invalid 'action' argument.");

        map<ActionToken*, StateHandler>::iterator it = _handlers.find(action);
        if (it != _handlers.end())
            it->second(data, result);
        else
            throw new UnknownActionException(action, GetToken());
    }
} }

File: Runtime/TypeScript/StateMachine/stateBase.ts
import token = require('token');
import stateManager = require('stateManager');
import events = require('events');
import transitionInfo = require('transitionInfo');

export class StateBase {

    private _handlers: {};
    private _token: token.StateToken;
    private _stateManager: stateManager.StateManager;

    constructor(token: token.StateToken) {
        if (!token)
            throw new Error('Invalid token argument.');

        this._token = token;
    }

    getToken(): token.StateToken {
        return this._token;
    }

    getStateManager(): stateManager.StateManager {
        return this._stateManager;
    }

    getContext(): any {
        if (!this._stateManager) {
            return null;
        }
        return this._stateManager.getContext();
    }

    onInitialized(): void { }
    onEnter(e: events.StateEnterEventArgs): void { }
    onExit(e: events.StateExitEventArgs): void { }

    /* internal */ __initialize(parent: stateManager.StateManager): void {
        this._stateManager = parent;
        if (typeof this.onInitialized === 'function') {
            this.onInitialized();
        }
    }

    registerHandler(action: token.ActionToken, handler: any): void {
        var key = action.getIdentifier().toString();
        if (this._handlers[key]) {
            throw new Error('Action already registered');
        }
        this._handlers[key] = handler;
    }

    /* internal */ __handle(action: token.ActionToken, data: any): transitionInfo.TransitionInfo {
        if (!action) {
            throw new Error('Invalid action argument.');
        }

        var key = action.getIdentifier().toString();
        if (!this._handlers[key]) {
            throw new Error('Unknown action. (' + action.toString() + ')');
        }

        var result = this._handlers[key](data);

        if (typeof result === 'TransitionInfo') {
            return result;
        } else if (typeof result === 'StateToken') {
            var ti = new transitionInfo.TransitionInfo();
            ti.targetStateData = data;
            ti.targetState = result;
            return ti;
        } else {
            throw new Error('Invalid handler result');
        }
    }
}
